fendermint/testing/materializer/src/docker/node.rs

```diff
@@ -0,0 +1,566 @@
+            fendermint,
+            cometbft,
+            ethapi,
+            port_range,
+            path: node_dir,
+        })
+    }
+}
+
+/// Create a container name from a node name and a logical container name, e.g. "cometbft"
+/// in a way that we can use it as a hostname without being too long.
+///
+/// It consists of `{node-id}-{container}-{hash(node-name)}`,
+/// e.g. "node-12-cometbft-a1b2c3"
+fn container_name(node_name: &NodeName, container: &str) -> String {
+    let node_id = node_name
+        .path()
+        .file_stem()
+        .unwrap_or_default()
+        .to_string_lossy()
+        .to_string();
+    let hash = ResourceHash::digest(node_name.path().to_string_lossy().to_string());
+    let hash = hash.to_string();
+    let hash = &hash.as_str()[..6];
+    format!("{node_id}-{container}-{}", hash)
+}
+
+struct DockerRunner<'a> {
+    dh: &'a DockerWithDropHandle,
+    node_name: NodeName,
+    image: String,
+    user: u32,
+    volumes: Volumes,
+}
+
+impl<'a> DockerRunner<'a> {
+    pub fn new(
+        dh: &'a DockerWithDropHandle,
+        node_name: &NodeName,
+        image: &str,
+        user: u32,
+        volumes: Volumes,
+    ) -> Self {
+        Self {
+            dh,
+            node_name: node_name.clone(),
+            image: image.to_string(),
+            user,
+            volumes,
+        }
+    }
+
+    fn docker(&self) -> &Docker {
+        &self.dh.docker
+    }
+
+    // Tag containers with resource names.
+    fn labels(&self) -> HashMap<String, String> {
+        [
+            ("testnet", self.node_name.testnet().path()),
+            ("node", self.node_name.path()),
+        ]
+        .into_iter()
+        .map(|(n, p)| (n.to_string(), p.to_string_lossy().to_string()))
+        .collect()
+    }
+
+    /// Run a short lived container.
+    pub async fn run_cmd(&self, cmd: &str) -> anyhow::Result<()> {
+        let config = Config {
+            image: Some(self.image.clone()),
+            user: Some(self.user.to_string()),
+            cmd: Some(vec![cmd.to_string()]),
+            attach_stderr: Some(true),
+            attach_stdout: Some(true),
+            labels: Some(self.labels()),
+            host_config: Some(HostConfig {
+                auto_remove: Some(true),
+                init: Some(true),
+                binds: Some(
+                    self.volumes
+                        .iter()
+                        .map(|(h, c)| format!("{}:{c}", h.to_string_lossy()))
+                        .collect(),
+                ),
+                ..Default::default()
+            }),
+            ..Default::default()
+        };
+
+        let id = self
+            .docker()
+            .create_container::<&str, _>(None, config)
+            .await
+            .context("failed to create container")?
+            .id;
+
+        self.docker()
+            .start_container::<&str>(&id, None)
+            .await
+            .context("failed to start container")?;
```

Comment:
In my experience in the next PR, it can fail if we misconfigure docker somehow, but having the failed container stick around means we can still inspect it to see what was wrong with it. It's easy to remove them with `docker system prune`, but I added a step to the tests to remove all previous containers labelled with the testnet.